File: apps/visualizer/backend/api/perspectives.py
# image_scores JSON field keys (library DB)
# image_key, image_type, perspective_slug, score, rationale, model_used, prompt_version, scored_at, is_current, repaired_from_malformed
"""REST API for the library ``perspectives`` registry.

List responses (``GET /api/perspectives``) expose these fields per row for UIs and
future catalog consumers: ``id``, ``slug``, ``display_name``, ``description``,
``active``, ``source_filename``, ``updated_at`` (no ``prompt_markdown`` on the list).

``image_scores`` rows (same library DB) use the JSON-friendly keys listed in the
header comment above when exposing score history over HTTP in later phases.
"""
from __future__ import annotations

import os
import re
from pathlib import Path

from flask import Blueprint, jsonify, request
from utils.db import with_db
from utils.responses import error_bad_request, error_not_found

from lightroom_tagger.core.database import (
    delete_perspective,
    get_perspective_by_slug,
    insert_perspective,
    list_perspectives,
    update_perspective,
)

bp = Blueprint("perspectives", __name__)

_REPO_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "..", ".."))

_SLUG_RE = re.compile(r"^[a-z][a-z0-9_]{0,63}$")
_SOURCE_FILENAME_RE = re.compile(r"^[a-zA-Z0-9_\-]+\.md$")
_MAX_PROMPT_MARKDOWN_BYTES = 256 * 1024


def _prompt_too_large(prompt_markdown: str) -> bool:
    return len(prompt_markdown.encode("utf-8")) > _MAX_PROMPT_MARKDOWN_BYTES


def _row_list_item(row: dict) -> dict:
    return {
        "id": row["id"],
        "slug": row["slug"],
        "display_name": row["display_name"],
        "description": row["description"],
        "active": bool(row["active"]),
        "source_filename": row.get("source_filename"),
        "updated_at": row.get("updated_at"),
    }


def _row_detail(row: dict) -> dict:
    out = _row_list_item(row)
    out["prompt_markdown"] = row["prompt_markdown"]
    out["created_at"] = row.get("created_at")
    return out


@bp.route("/", methods=["GET"])
@with_db
def list_perspectives_route(db):
    """List perspectives; optional ``active_only=true`` query."""
    active_only = request.args.get("active_only", "").lower() == "true"
    rows = list_perspectives(db, active_only=active_only)
    return jsonify([_row_list_item(r) for r in rows])


@bp.route("/<slug>", methods=["GET"])
@with_db
def get_perspective_route(db, slug: str):
    row = get_perspective_by_slug(db, slug)
    if not row:
        return error_not_found("resource")
    return jsonify(_row_detail(row))


@bp.route("/", methods=["POST"])
@with_db
def create_perspective_route(db):
    data = request.get_json(silent=True)
    if not data or not isinstance(data, dict):
        return error_bad_request("JSON body required")

    slug = data.get("slug")
    display_name = data.get("display_name")
    prompt_markdown = data.get("prompt_markdown")
    if not isinstance(slug, str) or not isinstance(display_name, str) or not isinstance(
        prompt_markdown, str
    ):
        return error_bad_request("slug, display_name, and prompt_markdown are required strings")
    if not _SLUG_RE.match(slug):
        return error_bad_request("invalid slug")
    if _prompt_too_large(prompt_markdown):
        return error_bad_request("prompt too large")

    description = data.get("description", "")
    if description is not None and not isinstance(description, str):
        return error_bad_request("description must be a string")
    description = description or ""

    active = data.get("active", True)
    if not isinstance(active, bool):
        return error_bad_request("active must be a boolean")

    if get_perspective_by_slug(db, slug):
        return error_bad_request("slug already exists")

    try:
        insert_perspective(
            db,
            slug=slug,
            display_name=display_name,
            prompt_markdown=prompt_markdown,
            description=description,
            active=active,
            source_filename=None,
        )
        db.commit()
    except Exception as e:
        db.rollback()
        return error_bad_request(str(e))

    created = get_perspective_by_slug(db, slug)
    assert created is not None
    return jsonify(_row_detail(created)), 201


@bp.route("/<slug>", methods=["PUT"])
@with_db
def update_perspective_route(db, slug: str):
    if not get_perspective_by_slug(db, slug):
        return error_not_found("resource")

    data = request.get_json(silent=True)
    if not data or not isinstance(data, dict):
        return error_bad_request("JSON body required")

    display_name = data.get("display_name") if "display_name" in data else None
    description = data.get("description") if "description" in data else None
    prompt_markdown = data.get("prompt_markdown") if "prompt_markdown" in data else None
    active = data.get("active") if "active" in data else None

    if (
        display_name is None
        and description is None
        and prompt_markdown is None
        and active is None
    ):
        return error_bad_request("at least one field required")

    if display_name is not None and not isinstance(display_name, str):
        return error_bad_request("display_name must be a string")
    if description is not None and not isinstance(description, str):
        return error_bad_request("description must be a string")
    if prompt_markdown is not None:
        if not isinstance(prompt_markdown, str):
            return error_bad_request("prompt_markdown must be a string")
        if _prompt_too_large(prompt_markdown):
            return error_bad_request("prompt too large")
    if active is not None and not isinstance(active, bool):
        return error_bad_request("active must be a boolean")

    updated = update_perspective(
        db,
        slug,
        display_name=display_name,
        description=description,
        prompt_markdown=prompt_markdown,
        active=active,
    )
    if not updated:
        return error_bad_request("no valid fields to update")
    db.commit()

    row = get_perspective_by_slug(db, slug)
    assert row is not None
    return jsonify(_row_detail(row))


@bp.route("/<slug>", methods=["DELETE"])
@with_db
def delete_perspective_route(db, slug: str):
    if not delete_perspective(db, slug):
        return error_not_found("resource")
    db.commit()
    return "", 204


@bp.route("/<slug>/reset-default", methods=["POST"])
@with_db
def reset_perspective_default_route(db, slug: str):
    row = get_perspective_by_slug(db, slug)
    if not row:
        return error_not_found("resource")

    source_filename = row.get("source_filename")
    if source_filename:
        if not isinstance(source_filename, str) or not _SOURCE_FILENAME_RE.match(source_filename):
            return error_bad_request("invalid source_filename")
        filename = source_filename
    else:
        filename = f"{slug}.md"

    path = (Path(_REPO_ROOT) / "prompts" / "perspectives" / filename).resolve()
    base = (Path(_REPO_ROOT) / "prompts" / "perspectives").resolve()
    try:
        path.relative_to(base)
    except ValueError:
        return error_bad_request("invalid source_filename")

    if not path.is_file():
        return jsonify({"error": "no default file"}), 404

    try:
        text = path.read_text(encoding="utf-8")
    except OSError as e:
        return error_bad_request(str(e))

    if _prompt_too_large(text):
        return error_bad_request("prompt too large")

    update_perspective(db, slug, prompt_markdown=text)
    db.commit()
    refreshed = get_perspective_by_slug(db, slug)
    assert refreshed is not None
    return jsonify(_row_detail(refreshed))

File: apps/visualizer/backend/src/api/perspectives.ts
/**
 * REST API for the library `perspectives` registry.
 *
 * List responses expose `id`, `slug`, `display_name`, `description`, `active`,
 * `optional`, `source_filename`, `updated_at` — no `prompt_markdown` on the list.
 * `optional` is read-only: it is derived from the markdown marker on write
 * (ADR-0012), never accepted from a body.
 */
import { createRoute, z } from '@hono/zod-openapi';
import { readFileSync, statSync } from 'node:fs';
import { join, resolve } from 'node:path';
import { REPO_ROOT } from '../config.js';
import {
  deletePerspective,
  getPerspectiveBySlug,
  insertPerspective,
  listPerspectives,
  updatePerspective,
  type PerspectiveRow,
} from '../db/library/scores.js';
import { libraryDb, type LibraryEnv } from './library-db.js';
import { isValidPerspectiveSlug } from '../utils/perspective-slug.js';
import { createOpenApiApp } from './openapi.js';
import { jsonBody, redirectToTrailingSlash, withValidationError } from './route-helpers.js';
import { ErrorBody } from './schemas/errors.js';
import {
  PerspectiveDetail,
  PerspectiveListResponse,
  PerspectiveSummary,
} from './schemas/perspectives.js';

export const perspectivesRoutes = createOpenApiApp<LibraryEnv>();

// This group writes (create/update/delete/reset), so the connection is writable.
perspectivesRoutes.use('/perspectives/*', libraryDb({ write: true }));

// Bare form redirects with 308 rather than 404ing.
redirectToTrailingSlash(perspectivesRoutes, '/perspectives');

/**
 * Route paths include the `/perspectives` prefix because the canonical URL is
 * `/api/perspectives/` with a trailing slash — mounting at `/api/perspectives` with
 * a bare `'/'` child would collapse to `/api/perspectives` without it.
 */

const SOURCE_FILENAME_RE = /^[a-zA-Z0-9_-]+\.md$/;
const MAX_PROMPT_MARKDOWN_BYTES = 256 * 1024;

function promptTooLarge(promptMarkdown: string): boolean {
  return Buffer.byteLength(promptMarkdown, 'utf8') > MAX_PROMPT_MARKDOWN_BYTES;
}

function rowListItem(row: PerspectiveRow): z.infer<typeof PerspectiveSummary> {
  return {
    id: row.id,
    slug: row.slug,
    display_name: row.display_name,
    description: row.description,
    active: Boolean(row.active),
    optional: Boolean(row.optional),
    source_filename: row.source_filename,
    updated_at: row.updated_at,
  };
}

function rowDetail(row: PerspectiveRow): z.infer<typeof PerspectiveDetail> {
  return {
    ...rowListItem(row),
    prompt_markdown: row.prompt_markdown,
    created_at: row.created_at,
  };
}

// --- list -------------------------------------------------------------------

const listRoute = createRoute({
  method: 'get',
  path: '/perspectives/',
  tags: ['perspectives'],
  responses: withValidationError({
    200: { description: 'Perspectives', content: jsonBody(PerspectiveListResponse) },
  }),
});

perspectivesRoutes.openapi(listRoute, (c) => {
  const activeOnly = (c.req.query('active_only') ?? '').toLowerCase() === 'true';
  const rows = listPerspectives(c.get('libraryDb'), { activeOnly });
  return c.json(rows.map(rowListItem), 200);
});

// --- detail -----------------------------------------------------------------

const detailRoute = createRoute({
  method: 'get',
  path: '/perspectives/{slug}',
  tags: ['perspectives'],
  request: { params: z.object({ slug: z.string() }) },
  responses: withValidationError({
    200: { description: 'One perspective', content: jsonBody(PerspectiveDetail) },
    404: { description: 'Not found', content: jsonBody(ErrorBody) },
  }),
});

perspectivesRoutes.openapi(detailRoute, (c) => {
  const row = getPerspectiveBySlug(c.get('libraryDb'), c.req.valid('param').slug);
  if (!row) return c.json({ error: 'resource not found' }, 404);
  return c.json(rowDetail(row), 200);
});

// --- create -----------------------------------------------------------------

const createPerspectiveRoute = createRoute({
  method: 'post',
  path: '/perspectives/',
  tags: ['perspectives'],
  responses: withValidationError({
    201: { description: 'Created', content: jsonBody(PerspectiveDetail) },
    400: { description: 'Invalid request', content: jsonBody(ErrorBody) },
  }),
});

/**
 * Body validation is hand-rolled to preserve the exact error strings the frontend
 * and contract tests assert on.
 */
perspectivesRoutes.openapi(createPerspectiveRoute, async (c) => {
  let data: unknown;
  try {
    data = await c.req.json();
  } catch {
    data = null;
  }
  if (!data || typeof data !== 'object' || Array.isArray(data)) {
    return c.json({ error: 'JSON body required' }, 400);
  }
  const body = data as Record<string, unknown>;

  const { slug, display_name: displayName, prompt_markdown: promptMarkdown } = body;
  if (
    typeof slug !== 'string' ||
    typeof displayName !== 'string' ||
    typeof promptMarkdown !== 'string'
  ) {
    return c.json(
      { error: 'slug, display_name, and prompt_markdown are required strings' },
      400,
    );
  }
  if (!isValidPerspectiveSlug(slug)) return c.json({ error: 'invalid slug' }, 400);
  if (promptTooLarge(promptMarkdown)) return c.json({ error: 'prompt too large' }, 400);

  const rawDescription = 'description' in body ? body.description : '';
  if (rawDescription != null && typeof rawDescription !== 'string') {
    return c.json({ error: 'description must be a string' }, 400);
  }
  const description = (rawDescription as string | null) || '';

  const active = 'active' in body ? body.active : true;
  if (typeof active !== 'boolean') return c.json({ error: 'active must be a boolean' }, 400);

  const db = c.get('libraryDb');
  if (getPerspectiveBySlug(db, slug)) return c.json({ error: 'slug already exists' }, 400);

  try {
    insertPerspective(db, {
      slug,
      displayName,
      promptMarkdown,
      description,
      active,
      sourceFilename: null,
    });
  } catch (e) {
    return c.json({ error: e instanceof Error ? e.message : String(e) }, 400);
  }

  const created = getPerspectiveBySlug(db, slug);
  // Thrown, not returned: the OpenAPI contract documents no 500 for this route.
  // Read-after-write on the connection that just wrote the row.
  if (!created) throw new Error('perspective vanished after insert');
  return c.json(rowDetail(created), 201);
});

// --- update -----------------------------------------------------------------

const updateRoute = createRoute({
  method: 'put',
  path: '/perspectives/{slug}',
  tags: ['perspectives'],
  request: { params: z.object({ slug: z.string() }) },
  responses: withValidationError({
    200: { description: 'Updated', content: jsonBody(PerspectiveDetail) },
    400: { description: 'Invalid request', content: jsonBody(ErrorBody) },
    404: { description: 'Not found', content: jsonBody(ErrorBody) },
  }),
});

perspectivesRoutes.openapi(updateRoute, async (c) => {
  const slug = c.req.valid('param').slug;
  const db = c.get('libraryDb');
  if (!getPerspectiveBySlug(db, slug)) return c.json({ error: 'resource not found' }, 404);

  let data: unknown;
  try {
    data = await c.req.json();
  } catch {
    data = null;
  }
  if (!data || typeof data !== 'object' || Array.isArray(data)) {
    return c.json({ error: 'JSON body required' }, 400);
  }
  const body = data as Record<string, unknown>;

  // Absent field means "leave alone", not falsy.
  const displayName = 'display_name' in body ? body.display_name : null;
  const description = 'description' in body ? body.description : null;
  const promptMarkdown = 'prompt_markdown' in body ? body.prompt_markdown : null;
  const active = 'active' in body ? body.active : null;

  if (displayName == null && description == null && promptMarkdown == null && active == null) {
    return c.json({ error: 'at least one field required' }, 400);
  }
  if (displayName != null && typeof displayName !== 'string') {
    return c.json({ error: 'display_name must be a string' }, 400);
  }
  if (description != null && typeof description !== 'string') {
    return c.json({ error: 'description must be a string' }, 400);
  }
  if (promptMarkdown != null) {
    if (typeof promptMarkdown !== 'string') {
      return c.json({ error: 'prompt_markdown must be a string' }, 400);
    }
    if (promptTooLarge(promptMarkdown)) return c.json({ error: 'prompt too large' }, 400);
  }
  if (active != null && typeof active !== 'boolean') {
    return c.json({ error: 'active must be a boolean' }, 400);
  }

  const updated = updatePerspective(db, slug, {
    displayName: displayName as string | null,
    description: description as string | null,
    promptMarkdown: promptMarkdown as string | null,
    active: active as boolean | null,
  });
  if (!updated) return c.json({ error: 'no valid fields to update' }, 400);

  const row = getPerspectiveBySlug(db, slug);
  if (!row) throw new Error('perspective vanished after update');
  return c.json(rowDetail(row), 200);
});

// --- delete -----------------------------------------------------------------

const deleteRoute = createRoute({
  method: 'delete',
  path: '/perspectives/{slug}',
  tags: ['perspectives'],
  request: { params: z.object({ slug: z.string() }) },
  responses: withValidationError({
    204: { description: 'Deleted' },
    404: { description: 'Not found', content: jsonBody(ErrorBody) },
  }),
});

perspectivesRoutes.openapi(deleteRoute, (c) => {
  if (!deletePerspective(c.get('libraryDb'), c.req.valid('param').slug)) {
    return c.json({ error: 'resource not found' }, 404);
  }
  return c.body(null, 204);
});

// --- reset to default -------------------------------------------------------

const resetRoute = createRoute({
  method: 'post',
  path: '/perspectives/{slug}/reset-default',
  tags: ['perspectives'],
  request: { params: z.object({ slug: z.string() }) },
  responses: withValidationError({
    200: { description: 'Reset to the on-disk default', content: jsonBody(PerspectiveDetail) },
    400: { description: 'Invalid request', content: jsonBody(ErrorBody) },
    404: { description: 'Not found', content: jsonBody(ErrorBody) },
  }),
});

perspectivesRoutes.openapi(resetRoute, (c) => {
  const slug = c.req.valid('param').slug;
  const db = c.get('libraryDb');
  const row = getPerspectiveBySlug(db, slug);
  if (!row) return c.json({ error: 'resource not found' }, 404);

  const sourceFilename = row.source_filename;
  let filename: string;
  if (sourceFilename) {
    if (!SOURCE_FILENAME_RE.test(sourceFilename)) {
      return c.json({ error: 'invalid source_filename' }, 400);
    }
    filename = sourceFilename;
  } else {
    filename = `${slug}.md`;
  }

  // Containment check before reading: `source_filename` comes from the database,
  // and the regex above is the only thing standing between it and an arbitrary
  // read. Resolve and confirm the result is still inside the prompts directory.
  const base = resolve(join(REPO_ROOT, 'prompts', 'perspectives'));
  const path = resolve(join(base, filename));
  if (path !== base && !path.startsWith(base + '/')) {
    return c.json({ error: 'invalid source_filename' }, 400);
  }

  try {
    if (!statSync(path).isFile()) return c.json({ error: 'no default file' }, 404);
  } catch {
    return c.json({ error: 'no default file' }, 404);
  }

  let text: string;
  try {
    text = readFileSync(path, 'utf8');
  } catch (e) {
    return c.json({ error: e instanceof Error ? e.message : String(e) }, 400);
  }

  if (promptTooLarge(text)) return c.json({ error: 'prompt too large' }, 400);

  updatePerspective(db, slug, { promptMarkdown: text });
  const refreshed = getPerspectiveBySlug(db, slug);
  if (!refreshed) throw new Error('perspective vanished after reset');
  return c.json(rowDetail(refreshed), 200);
});
